src/codegen/statements.rs

```diff
@@ -63,21 +64,42 @@ pub(crate) fn statement(
                     vartab,
                     opt,
                 };
-
                 //If we remove the assignment, we must keep expressions that have side effects
                 init.recurse(&mut params, process_side_effects_expressions);
                 return;
             }
 
             let expr = expression(init, cfg, contract_no, Some(func), ns, vartab, opt);
+
             cfg.add(
                 vartab,
                 Instr::Set {
                     loc: *loc,
                     res: *pos,
-                    expr,
+                    expr: expr.clone(),
                 },
             );
+
+            // Lets check if the declaration is a declaration of a dynamic array
+            if let Expression::AllocDynamicArray(_, _, size, _) = expr {
+                let temp_res = vartab.temp_name("array_size", &Type::Uint(32));
+                cfg.add(
+                    vartab,
+                    Instr::Set {
+                        loc: *loc,
+                        res: temp_res,
+                        expr: *size,
```

Comment:
We have a problem here. The `size` expression is added twice to the cfg, duplicating code.

Say we have:
```
int[] foo = new int[]((x * y + 2)^3);
```
Now the cfg will have evaluate `(x * y + 2)^3` twice.

For the case that `expr = Expression::AllocDynamicArray(_, _, size, _)`, create a temporary variable like you are doing here, and then replace `expr` with `Expression::AllocDynamicArray(.., .., Box::new(Expression::Variable(loc, ty, temp_res), ..);` (with all the other fields copied) before adding it to the cfg.

That avoid the `.clone()` above too.

Reply:
This set instruction sets the temp variable to the size expression of the AllocDynamicArray expression (variable/numberliteral etc...), why should be expr replaced with AllocDynamicArray?